### PR TITLE
Add CUDA Scan operator.

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Add CUDA Scan operator.
Uses CPU implementation for logic.
Added some device specific functors for handling when data needs to be manipulated on a different device.
Added ability to override the materialization logic in the OrtValue slicer so DML can plugin their handling.

**Motivation and Context**
Avoids copy to CPU for all Scan inputs and outputs if they are currently on or needed on GPU. Also means that if the subgraph consumes the data on GPU it can stay there the whole time.

### CODE REVIEW

File: LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
```
Comment (KeDengMS) [inline LICENSE.txt]: unrelated change?

**Comment by KeDengMS** [comment]
> 

unrelated?

---
Refers to: this_build:3 in fa150c1. [](commit_id = fa150c195ca9b0b99b0c6c59c5690c9def7364c2, deletion_comment = False)

File: onnxruntime/core/providers/cuda/controlflow/scan.cc [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  auto status = onnxruntime::Scan<9>::Compute(ctx);
+  return status;
+}
+
+ONNX_OPERATOR_VERSIONED_KERNEL_EX(Scan,
+                                  kOnnxDomain,
+                                  8, 8,
+                                  kCudaExecutionProvider,
+                                  KernelDefBuilder()
+                                      .InputMemoryType<OrtMemTypeCPUInput>(0)  // 'sequence_lens' needs to be on CPU
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/controlflow/scan.cc]: >sequence_lens [](start = 83, length = 13)

Putting Scan op in CUDA device has some implications about memory placement for its inputs and outputs. I think a more general way might be to look at the device info in CPU Scan implementation, for the inputs and outputs of Scan node and node placement in its subgraph? Otherwise, Scan9+ may still have suboptimal perf from copies between CPU and GPU.

File: LICENSE.txt (hunk 1 above)
Comment (skottmckay) [reply LICENSE.txt]: Yeah - not sure where the change came from and didn't intend for that to be picked up. Will remove.

File: onnxruntime/core/providers/cuda/controlflow/scan.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cuda/controlflow/scan.cc]: I initially looked at what would be required to check the device info for the Scan inputs/outputs and subgraph inputs/outputs and avoid adding memcpy nodes where possible, but the amount of change to do that is extremely large. We'd have to move around a lot of the initial processing as currently we fully process the main graph followed by subgraphs. Instead you'd have to split major parts of the initialization up a lot differently, and go bottom up so that you have full knowledge of the subgraph partitioning when processing the parent graph. It would also make operator specific logic bleed into general setup as you would need to know which specific inputs of Scan/Loop/If needed to be on CPU and which didn't. There are a few weeks of work in making that sort of change. 

Doing all that would only save you a copy if there was a mismatch between the location of the Scan inputs/outputs and the location of the subgraph inputs/outputs. If we make an assumption that we want minimal gap between the operators available on CUDA and CPU (where possible), that could make such a mismatch rare. 

So whilst it would be more general, I thought the cost/benefit of taking this approach was a clearer win and the changes required are a lot cleaner and better separated. If we still see performance issues with this approach that may validate doing things differently.
